libraries/extractor/src/main/java/androidx/media3/extractor/ts/DtsReader.java

```diff
@@ -95,20 +133,66 @@ public void packetStarted(long pesTimeUs, @TsPayloadReader.Flags int flags) {
   }
 
   @Override
-  public void consume(ParsableByteArray data) {
+  public void consume(ParsableByteArray data) throws ParserException {
     Assertions.checkStateNotNull(output); // Asserts that createTracks has been called.
     while (data.bytesLeft() > 0) {
       switch (state) {
         case STATE_FINDING_SYNC:
           if (skipToNextSync(data)) {
-            state = STATE_READING_HEADER;
+            if (isFtocSync) {
+              uhdInSync = true;
+              state = STATE_FINDING_UHD_HEADER_SIZE;
+            } else if (isFtocNonSync) {
+              state = STATE_FINDING_UHD_HEADER_SIZE;
+            } else if (isCoreSync) {
+              state = STATE_READING_CORE_HEADER;
+            } else {
+              state = STATE_FINDING_EXTSS_HEADER_SIZE;
+            }
           }
           break;
-        case STATE_READING_HEADER:
-          if (continueRead(data, headerScratchBytes.getData(), HEADER_SIZE)) {
-            parseHeader();
+        case STATE_READING_CORE_HEADER:
+          if (continueRead(data, headerScratchBytes.getData(), CORE_HEADER_SIZE)) {
+            parseCoreHeader();
             headerScratchBytes.setPosition(0);
-            output.sampleData(headerScratchBytes, HEADER_SIZE);
+            output.sampleData(headerScratchBytes, CORE_HEADER_SIZE);
+            state = STATE_READING_SAMPLE;
+          }
+          break;
+        case STATE_FINDING_EXTSS_HEADER_SIZE:
+          // Read enough bytes to parse the header size information.
+          if (continueRead(data, headerScratchBytes.getData(), /* targetLength= */ 10)) {
+            extensionSubstreamHeaderSize =
+                DtsUtil.parseDtsHdHeaderSize(headerScratchBytes.getData());
+            state = STATE_READING_EXTSS_HEADER;
+          }
+          break;
+        case STATE_READING_EXTSS_HEADER:
+          if (continueRead(data, headerScratchBytes.getData(), extensionSubstreamHeaderSize)) {
+            parseExtensionSubstreamHeader();
+            headerScratchBytes.setPosition(0);
+            output.sampleData(headerScratchBytes, extensionSubstreamHeaderSize);
+            state = STATE_READING_SAMPLE;
+          }
+          break;
+        case STATE_FINDING_UHD_HEADER_SIZE:
+          // Read enough bytes to parse the header size information.
+          if (continueRead(data, headerScratchBytes.getData(), /* targetLength= */ 6)) {
+            uhdHeaderSize = DtsUtil.parseDtsUhdHeaderSize(headerScratchBytes.getData());
+            // Adjust the array read position if data read is more than the actual header size.
```

Comment:
I believe this is no longer required as we always start with a SYNC frame. If you'll confirm, I can remove this.

Reply:
The header size extraction needs to be performed both in the FTOC sync frame and FTOC non-sync frames.
 
Every UHD frame contains an FTOC(Frame Table of Contents) component. The main component of the FTOC is the Sync Word (the first 4 bytes of the FTOC), which indicates whether the frame is a sync frame or a non-sync frame. 

In the sync frame(FTOC sync frame), the decoder does not need any information from previous frames to produce output. All parameters necessary to unpack metadata and audio chunks can be found within the payload of this sync
frame.

A non-sync frame(FTOC non-sync frame) permits both metadata chunks and audio chunks to minimize payload size by only sending parameters that have changed in value since the previous frame or sync frame. All parameters that are not re-transmitted are assumed to maintain their previous value. Any value or set of values may be updated in a non-sync frame.